language/diem-vm/parallel-executor/src/task.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use mvhashmap::{MVHashMapView, Version};
+use std::{fmt::Debug, hash::Hash};
+
+/// The execution result of a transaction
+#[derive(Debug)]
+pub enum ExecutionStatus<T, E> {
+    /// Transaction was executed successfully.
+    Success(T),
+    /// Transaction hit a none recoverable error during execution, halt the execution and propagate
+    /// the error back to the caller.
+    Abort(E),
+    /// Transaction was executed successfully, but will skip the execution of the trailing
+    /// transactions in the list
+    SkipRest(T),
+    /// Transaction has an unexpected read dependency that is blocked by `Version` transaction.
+    /// Put the transaction back to the scheduler.
+    Retry(/*blocked_by*/ Version),
+}
+
+/// Trait that defines a transaction that could be parallel executed by the scheduler. Each
+/// transaction will write to a key value storage as their side effect.
+pub trait Transaction: Clone + Sync + Send + 'static {
+    type Key: PartialOrd + Send + Sync + Clone + Hash + Eq;
+    type Value: Send + Sync;
+}
+
+/// Trait for inferencing the read and write set of a transaction.
+pub trait TransactionInferencer: Sync {
+    /// Type of transaction and its associated key.
+    type T;
```

Comment:
I agree with Todd here. I think even though it might be a bit more painful, the clarity that it brings might be worth it. Also, I'm not fully up on my associated type foo, but if `type T: Transaction` can't you just do something like `Self::T::Key`?